### PR TITLE
added unit tests

### PR DESCRIPTION
**Pull request by madhu-md**

### CODE REVIEW

File: chatbot_train/test_train.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        data = self.test_cfg["data6"]
+        epochs = self.test_cfg["epochs1"]
+        # test with empty data
+        self.assertFalse(train_run(data,epochs))
+
+    def test_null(self):
+        # test with null data
+        data = self.test_cfg["data7"]
+        epochs = self.test_cfg["epochs1"]
+        self.assertFalse(train_run(data,epochs))
```
Comment (orionsBeltWest) [inline chatbot_train/test_train.py]: It looks that train_run returns bool is related to the validity of its arguments: data and epochs.  and not related to actual training runs. 
These are valid test cases. 
However, we need to test if the actual train results are within expectation:
```python:
        print(f'\nTraining Loss: {train_loss:.3f}')
        print(f'\nTraining Accuracy: {train_accuracy:.3f}')
        print(f'\nTest Accuracy: {test_accuracy:.3f}') 
```

File: chatbot_inference/test_predict.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        self.empty_data = self.test_cfg["empty_data"][0]
+        self.list_data = self.test_cfg["list_data"][0]
+        self.large_data = self.test_cfg["large_data"][0]
+        self.num_data = self.test_cfg["num_data"][0]
+
+    # test with data
+    def test_with_data(self):
+        data = self.data
+        print("----testing with data :", data)
+        self.assertTrue(get_response(data))
```
Comment (orionsBeltWest) [inline chatbot_inference/test_predict.py]: The return type of get_response is dict

```python: 
result = {"intent": intent, "response": result, "confidence": confidence}
```

hence I would expect using self.assertEqual or self.assertAlmostEqual for the confidence part?

**Review by orionsBeltWest — CHANGES_REQUESTED** [review]
The training test cases look good however, I would recommend adding several to testing the actual training results. Not just the arguments.  Probably just have the function return a tuple of these results. 

The predict test cases are using self.assertTrue for a dict return.  This need to change to self.assertEqual or it equivalent for examining disct values.

**Review by jedwards-habana — CHANGES_REQUESTED** [review]
All the yaml and py files will need to have the appropriate license added to the header (MIT). Workflow files should follow the convension, pr-<unit test name>.yaml but shouldn't have unittests on the end.

Also, all job names should be changed from 'build"

File: .github/workflows/inference_unittests.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  push:
+    branches:
+      - md-unittests
+    paths: 
+      - chatbot_inference/**
+      
+  pull_request:
+
+jobs:
+  build:
```
Comment (jedwards-habana) [inline .github/workflows/inference_unittests.yaml]: unittest

File: .github/workflows/train_unittests.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  push:
+    branches:
+      - md-unittests
+    paths:
+      - chatbot_train/**
+      
+  pull_request:
+
+jobs:
+  build:
```
Comment (jedwards-habana) [inline .github/workflows/train_unittests.yaml]: unittest

**Review by jedwards-habana — CHANGES_REQUESTED** [review]

File: .github/workflows/inference_unittests.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Setup Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: '3.9'
+      - name: test 
+        run: ls chatbot_inference
+        working-directory: ${{ env.GITHUB_WORKSPACE }}
```
Comment (jedwards-habana) [inline .github/workflows/inference_unittests.yaml]: Unneeded step. Please remove.

File: .github/workflows/inference_unittests.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+      - name: Install libs
+        run: |
+          #python -m pip install --upgrade pip
+          pip install -r chatbot_inference/requirements.txt
+        working-directory: ${{ env.GITHUB_WORKSPACE }}
+      - name: Unit test execution
```
Comment (jedwards-habana) [inline .github/workflows/inference_unittests.yaml]: I was thinking we can use the find -exec functionality to execute unit test files. First, we should come up with a common naming conventions for unit test tile. Something like unittest_*.py. I think you use test_* but lets change that to unittest_*. Then, the command to execute all the unit tests in a directory would be:
.
      - name: Execute unit tests
        run: |
          find ./ncf_inference -name unittest*.py -exec python {} \;
.
This way, any unit test file added to the directory automatically gets appended to the list.

**Review by jedwards-habana — CHANGES_REQUESTED** [review]

File: chatbot_train/test_train.py (hunk 1 above)
Comment (madhu-md) [reply chatbot_train/test_train.py]: As per your request, I have made few changes and tested with  train accurracy. Please check and let me know if  any modifications needed.

File: .github/workflows/inference_unittests.yaml (hunk 3 above)
Comment (madhu-md) [reply .github/workflows/inference_unittests.yaml]: I have updated all the pipelines as per the standards. Please take a look at it.

File: chatbot_train/test_train.py (hunk 1 above)
Comment (madhu-md) [reply chatbot_train/test_train.py]: @orionsBeltWest As discussed, I have made few changes and tested with train accurracy. Please check and let me know if any modifications needed.

**Review by jedwards-habana — APPROVED** [review]
This looks good and I will approve it. In the future, lets separate the unit tests from the pipelines when we do a pull request, however.

File: chatbot_train/unittest_train.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    def test_with_parms(self):
+        #setup data
+        data = self.test_cfg["data1"]
+        epochs = self.test_cfg["epochs1"]
+        #test with data and epochs values
+        train = train_run(data,epochs)[1]
+        try:
+           self.assertAlmostEqual(train,25,places=1)
+        except AssertionError:
```
Comment (orionsBeltWest) [inline chatbot_train/unittest_train.py]: let's us discuss this? why the test case is troughing an exception?

File: chatbot_train/unittest_train.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            pass
+        self.assertTrue(train_run(data,epochs))
+    
+    def test_negative_epoch(self):
+        #setup data
+        data = self.test_cfg["data2"]
+        epochs = self.test_cfg["epochs2"]
+        self.assertEqual(train_run(data,epochs),False)
```
Comment (orionsBeltWest) [inline chatbot_train/unittest_train.py]: train runs returns tuple ! 
return train_loss, train_accuracy, test_accuracy 

not bool ?

File: chatbot_inference/unittest_predict.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        data = self.data
+        print("----testing with data :", data)
+        #self.assertTrue(get_response(data))
+        self.assertGreater(get_response(data)['confidence'], 0.80 )
+
+    def test_without_data(self):
+        data = self.empty_data
+        # test without data 
+        print("----testing without data :", data)
+        self.assertTrue(get_response(data))
```
Comment (orionsBeltWest) [inline chatbot_inference/unittest_predict.py]: The return type is tuple.  Not a bool 
result = {"intent": intent, "response": result, "confidence": confidence}

Comment (orionsBeltWest) [reply chatbot_inference/unittest_predict.py]: let's discuss

**Review by orionsBeltWest — CHANGES_REQUESTED** [review]

File: chatbot_inference/unittest_predict.py (hunk 9 above)
Comment (madhu-md) [reply chatbot_inference/unittest_predict.py]: @orionsBeltWest I have cleaned up the unwanted data and removed the bool and excepton in test cases for the traina and inference. Please look into this let me know if any changes required.

File: chatbot_train/unittest_train.py [hunk 10]
```diff
@@ -21,56 +21,48 @@ def setUpClass(self):
...
         data = self.test_cfg["data1"]
         epochs = self.test_cfg["epochs1"]
         #test with data and epochs values
         train = train_run(data,epochs)[1]
-        try:
-           self.assertAlmostEqual(train,25,places=1)
-        except AssertionError:
-            pass
-        self.assertTrue(train_run(data,epochs))
+        self.assertNotAlmostEqual(train,0.0)
```
Comment (orionsBeltWest) [inline chatbot_train/unittest_train.py]: Two issues remaining:

Why comparing to 0.0? 

As discussed,  train_run returns a tuple:
``` python:
 return train_loss, train_accuracy, test_accuracy
```

Hence you will need to compare this tuple

**Review by orionsBeltWest — CHANGES_REQUESTED** [review]
predict is fine. 

unittest_train.py still having issues

File: chatbot_train/unittest_train.py (hunk 10 above)
Comment (madhu-md) [reply chatbot_train/unittest_train.py]: @orionsBeltWest  I have refactoring the code for train and made changes to unittest file. Please review.

**Review by orionsBeltWest — APPROVED** [review]

**Pull request merged**